Update PortAudio

dyfer opened the pull request:
<!-- Please see CONTRIBUTING.md for guidelines. -->

## Purpose and Motivation

<!-- If this fixes an open issue, link to it by writing "Fixes #555." -->

This PR removes our fork of PortAudio and adds official PortAudio repository from ~~Assembla~~ GitHub. Our CMakeLists is modified to accommodate the change.

## Types of changes

<!-- Delete lines that don't apply -->

- Bug fix (?)
- New feature (?)
- ~~Breaking change~~ hopefully not

## To-do list

<!-- Complete an item by checking it: [x]. Add new entries to track your progress -->

- [x] Code is tested
- [x] Documentation is updated
  - will update once we agree on other aspects of this PR
- [x] This PR is ready for review
- [ ] commit history is cleaned up

dvzrv commented:
@dyfer that's great news. Is the upstream commit current HEAD or the latest tag (sadly very old)?

dyfer commented:
@dvzrv thanks! I've put it at current HEAD.
I've only checked whether server boots and makes sound, it would be great for Windows users to put it through its paces to make sure it still performs well.

brianlheim commented on CMakeLists.txt:
```diff
@@ -84,11 +84,16 @@ if(WIN32)
+    # Call static library "portaudio_static" since this will also cause the inclusion of the required architecture suffix
+    set(PA_LIBNAME_ADD_SUFFIX ON CACHE BOOL "PortAudio static library suffix")
+    # disable PortAudio install target
+    set(PA_DISABLE_INSTALL ON CACHE BOOL "Disable PortAudio install")
```
i think these should go inside the `if(NOT MSYS)` guard. i thought you could get away with not caching these, but apparently not until cmake 3.13: https://cmake.org/cmake/help/latest/policy/CMP0077.html#policy:CMP0077

dyfer replied:
`set(PA_LIBRARY_TYPE "STATIC")` was previously before `if(NOT MSYS)`, that's why I left them there. Wouldn't want these options to be set for all Win32 builds?

brianlheim replied:
maybe i'm missing something... what effect do these settings have in MSYS builds?

dyfer replied:
When I moved things to the `external_libraries`'s CMakeLists, I removed MSYS checks. I don't see how they are needed... (?)

brianlheim commented on CMakeLists.txt:
```diff
@@ -84,11 +84,16 @@ if(WIN32)
     if(NOT MSYS)
-      set(PORTAUDIO_INCLUDE_DIRS "${CMAKE_SOURCE_DIR}/external_libraries/portaudio_sc_org/include" CACHE PATH "Portaudio include directory" FORCE)
-      add_subdirectory("${CMAKE_SOURCE_DIR}/external_libraries/portaudio_sc_org")
+      set(PORTAUDIO_INCLUDE_DIRS "${CMAKE_SOURCE_DIR}/external_libraries/portaudio/include" CACHE PATH "Portaudio include directory" FORCE)
```
do you happen to know if this line is strictly necessary? i would expect linking against the imported lib should bring in this include directory as a transitive property, but also see that portaudio's CMakeLists.txt uses this BUILD_INTERFACE property that i'm not familiar with. it would be nice to get rid of this line because we could then move this section into `external_libraries/CMakeLists.txt` where it would be more appropriate.

dyfer replied:
looking more into this, `PORTAUDIO_INCLUDE_DIRS` are set by `FindPortaudio`, if it is found. If this (and `PORTAUDIO_LIBRARIES`) are specified when running `FindPortaudio`, it assumes that PA has been found. Thus, we set these when building PA ourselves.

brianlheim commented on lang/CMakeLists.txt:
```diff
@@ -297,7 +297,7 @@ endif(ALSA_FOUND)
 if (AUDIOAPI STREQUAL portaudio)
     if(WIN32 AND NOT MSYS)
-        target_link_libraries(libsclang portaudio)
+        target_link_libraries(libsclang portaudio_static)
```
if, after https://github.com/supercollider/supercollider/pull/4925/files#diff-af3b638bc2a3e6c650974192a53c7291R96,  you `set(PORTAUDIO_LIBRARIES portaudio_static)`, it appears you could entirely eliminate this `if` and the one in server/scsynth/CMakeLists.txt

brianlheim commented on server/supernova/CMakeLists.txt:
```diff
@@ -162,7 +162,7 @@ elseif(AUDIOAPI STREQUAL portaudio)
     if(WIN32)
         target_compile_definitions(libsupernova PUBLIC PORTAUDIO_BACKEND -DPORTAUDIO_BACKEND)
-        target_link_libraries(libsupernova portaudio)
+        target_link_libraries(libsupernova portaudio_static)
```
here also, you could eliminate this if() -- `${PORTAUDIO_DEFINITIONS}` never expands to anything, and `target_compile_definitions(libsupernova PUBLIC PORTAUDIO_BACKEND -DPORTAUDIO_BACKEND)` is exactly equivalent to `target_compile_definitions(libsupernova PUBLIC PORTAUDIO_BACKEND)`

dyfer commented:
I'll be working to address the review soon.

@brianlheim what do you think about changing the logic for when to build using provided ("vendored"?) PA? How about, instead of doing that on windows only, we add a flag like `USE_VENDORED_PORTAUDIO` and set it by default to true on windows and false elsewhere? I'm thinking it might be useful to have the option to build portaudio in specific cases (RPi maybe). Since we're already shipping PA, why not allow users to compile it on other platforms if needed?

bagong commented:
I think fundamentally it's a good move... But let me bring forward two points - not with the expectation that they are met - but as input worth a consideration: Assembla is a terribly inaccessible site, I think you can't even make pull requests without a paid account. So what happened in the past is that people made suggestions on the PA mailing list - maybe even sent patches per email ;), and then one of the acted, or it . I think that is one of the main reasons why there are so many PA forks on Github. One might also consider a policy that all repos SC pulls in should "buffered" within the SC org. I mean getting vanilla PA currently just requires to switch to master, probably not even pull, in our fork here... 

The other idea might be to have a look at vcpkg. They have packages for PA, libsndfile, fftw3 and even readline (which might be worth a revisit). They have Qt too, but last time I looked webengine was missing. But getting all the small dependencies from a single source, built consistently with the same toolset as SC might be a nice mid- or long-term goal. I remember scepticism towards Homebrew back in the day, but I think nobody gives any alternative a thought these days. MS seems to be eager to make vcpkg the homebrew of the Windows world (they're even trying to go cross-platform).

dyfer commented:
Thanks for your input @bagong !

I don't know much about Assembla, but it sounds like these are difficulties in developing PA. How does that inform the choice of us using it as the submodule repo?

Re our portaudio fork: AFAICT the changes from the `supercollider` branch were pulled into `master` so it's no longer possible to just update master from the upstream. Maybe it would work if we had another branch named differently that would track `assembla/master`? I don't know git enough to consider that. (FWIW it seems that our changes to PA are no longer needed (neither for MSVC nor MinGW), considering additions in this PR).

As for hosting our own forks - it seemed to me that we are trying to avoid that. I think this translates to additional burden in updating/maintaining these... but I'm not sure if that's the primary reason. We certainly are _not_ doing that in general - even our fork of `nova_tt` is not used as a submodule, but it's the original Tim's repo instead.

`vcpkg` looks interesting, but I don't have any experience with it. That seems to be an idea worth considering indeed (especially if they provide libsndfile!). Do we know about its availability on AppVeyor (or even Travis)?

bagong commented:
Sure, @dyfer, unfortunately I don't have any more specific input except bringing up these broader issues. Pulling a submodule directly from the source - while it's unlikely - could face seeing the source disappear ;) And if a community is very irresponsive it might be necessary - while unwelcome - to make changes downstream... As to vpkg's availability on AppVeyor I am not informed aboout the current state either. I just wanted to drop the idea... Back in the day vcpkg didn't provide binaries, just a reliable environment. Not sure if that changed either...

brianlheim commented:
@bagong thanks for the suggestions, this is really useful!

i have a long comment here so i've added topic headers.

### mirroring dependency repositories

while i do see the benefit of creating our own mirror for dependencies, it is also an extra burden for project maintainers, for very little real gain in my opinion. unless we want to take on the burden of maintaining an entirely new project alongside supercollider, patches still need to be submitted upstream, but now two sets of repo updates need to take place to get it into this repo. this discourages frequent updates and is also the first step toward our fork diverging from upstream. and if that happens, we are stuck keeping that fork alive for the rest of eternity, because it contains commits that upstream does not, and which are pointed to by this repo. that is a truly unfortunate situation in my opinion. i strongly feel we should not be adding to this entropy, even if it means waiting for an unresponsive or unwelcoming upstream to patch and release.

if a patch becomes necessary, as a last resort, i think it is better to automate that using CMake and git patches inside our own repository. then we can continue to pull in upstream changes while applying our fixes on top.

### upstream dependencies moving

if the original source moves, even if that happens abruptly, we can update with `git submodule set-url`, and everyone who has currently forked supercollider can update with `git submodule sync`. in the very rare event that this happens, communicating it publicly via the forum, mailing lists, or even some sort of check and notification in the build system is an okay solution as far as i'm concerned.

from a maintainer's perspective, my understanding is that, as long as the new fork maintains the same git history as the old, there is no issue going back to commits in the parent repo prior to the submodule change and running `git submodule update`. `git submodule update` checks out the commit independently of the url it has been sync'd or initialized with. in other words, you carry that new submodule URL with you even as you traverse old commits where `.gitmodules` contains a different URL for the submodule, until you run `git submodule sync` or `set-url`.

so, in a nutshell, i don't see any short term or long term problems caused by using an upstream repo directly, considering how rare migrations are.

### vcpkg

vcpkg is definitely worth another look based on what i'm seeing and what you said here. i searched around for what other people in C++ think of vcpkg, and based on [this reddit thread](https://www.reddit.com/r/cpp/comments/gch3i5/your_thoughts_on_vcpkg/), it seems they are still working on some basic core features, and given that we already have a workflow with submodules that works OK, i am somewhat hesitant to say let's go ahead and spend time switching everything over. it would be really helpful to know if other people have used vcpkg and what their experience was.

> They have Qt too, but last time I looked webengine was missing.

it appears to be present now - https://repology.org/project/qt/versions - but at 5.12.8, not 5.14. perhaps they are sticking to LTS intentionally. in any case, Qt is the least difficult part of the Windows build to configure, in my experience. if we can use vcpkg for fftw3, readline, libsndfile, and portaudio, that would be a huge simplification to our build process!

also, for reference i see that vcpkg is available on Appveyor build machines by default, but not in Travis's beta Windows support, where they do have chocolatey.

### regarding this PR

let's please keep this PR focused, and not let an important discussion happen where people are unlikely to find it easily. i would really urge to move any further conversation or proposals regarding vcpkg to an issue ticket (feel free to use the "reference in new issue" option on this comment). it is really not fair to @dyfer to have that discussion here, i think, as it creates the impression he is the sole person responsible for researching and resolving all conversations about vcpkg/submodules.

switching up submodules is no small change, and i do feel it would be good to take some time to consider our options. if in a couple weeks' time there is still no conclusive opinion about vcpkg, i think we should continue with this PR, since it is (1) in line with our existing practices, (2) brings in improvements in a dependency, possibly bugfixes, and (3) does contribute to the overall cleanliness of this project. @dyfer and @bagong, does that sound good to you?

brianlheim commented:
@dyfer given the above perhaps it would be best to not do any extra work and focus on the conversation about vcpkg first? then once that's resolved we can return to this PR, if that's all right.

dyfer commented:
Thanks sounds good @brianlheim, thanks for extensive comment and the summary. I don't mind if we look into using vcpkg first.

bagong commented:
@dyfer & @brianlheim  Sure! Good moves!

dyfer commented:
I think it would be good to have the option to use the latest PortAudio, i.e. not from a package manager. I updated this PR based on previous feedback.

dyfer commented:
After a discussion on a recent dev call I was investigating using CMake's `ExternalProject_Add` for adding PortAudio. Upon further investigation it doesn't seem like the best option: since using `ExternalProject` pushes PortAudio's configuration later, to when SC _building_ happens, it is difficult to set up linking properly (or at least I failed to do that). It seems that the recommended way to use `ExternalProject` is to have both main project (here SC) and the external library (here PA) as external projects. This seems like unnecessarily redoing a lot of CMake, for what might not even be a permanent dependency (if we decide to move away from building PA ourselves in the future).

Thus, I cleaned up the current `add_subdirectory` solution. The PR is ready for another review. Thanks!

dyfer commented:
After some more thought about this I realized we could put portaudio in a subdirectory. I think this makes disabling cmake flags and definitions cleaner (since they don't need to be re-set). It also makes it (marginally) cleaner with asiosdk IMO, which now needs to go into `external_libraries/portaudio` (instead of `external_libraries`). AppVeyor script is updated accordingly.

dyfer commented:
This PR is now updated to use the [Portaudio remote on Github](https://github.com/PortAudio/portaudio). Windows readme is also updated.

brianlheim commented on CMakeLists.txt:
```diff
@@ -231,6 +224,12 @@ option(SC_SYMLINK_CLASSLIB "Place a symlink of SCCLassLibrary instead of copying
+if(WIN32)
+    set(BUILD_PORTAUDIO ON)
+else()
+    mark_as_advanced(BUILD_PORTAUDIO)
```
this option should be SYSTEM_PORTAUDIO like the others

brianlheim commented on external_libraries/CMakeLists.txt:
```diff
@@ -6,6 +6,11 @@ if (NOT EXISTS ${CMAKE_CURRENT_SOURCE_DIR}/nova-simd/vec.hpp)
 
+if(BUILD_PORTAUDIO)
+  message(STATUS "PortAudio: building static library")
+  add_subdirectory("${CMAKE_SOURCE_DIR}/external_libraries/portaudio")
```
would also recommend 4 spaces here since the block above is also 4 and i do plan on reformatting this eventually, but up to you

brianlheim commented on external_libraries/portaudio/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+unset(CMAKE_C_FLAGS)
+unset(CMAKE_CXX_FLAGS)
+if(WIN32)
+  remove_definitions(-DWIN32_LEAN_AND_MEAN)
```
4 spaces (https://github.com/supercollider/supercollider/wiki/Code-style-guidelines#other-files)

brianlheim commented on external_libraries/portaudio/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+add_subdirectory("${CMAKE_SOURCE_DIR}/external_libraries/portaudio/portaudio_library")
+
+# when building PA ourselves, we need to specify PORTAUDIO_INCLUDE_DIRS and PORTAUDIO_LIBRARIES
+set(PORTAUDIO_INCLUDE_DIRS "${CMAKE_SOURCE_DIR}/external_libraries/portaudio/portaudio_library/include" CACHE PATH "Portaudio include directory" FORCE)
```
`${CMAKE_SOURCE_DIR}/external_libraries/portaudio/...` can be replaced with `${CMAKE_CURRENT_SOURCE_DIR}/...` throughout

brianlheim reviewed (changes requested):
thanks for updating!

dyfer commented:
@brianlheim I addressed review comments
As far as testing goes - I've built this and have run servers on one machine with both internal and external soundcard, putting some stress on it (~500 SinOsc synthdefs) with MME and ASIO drivers, for a relatively short time. The performance seems unchanged from the previous implementation, but due to nature of Windows sound and PortAudio this needs to be tested "in the wild" to really know if there are issues.
Also, PA builds not only on Windows, but also on macOS (I might have tried on Linux as well but can't remember ATM).

dyfer commented on .gitmodules:
```diff
@@ -23,3 +19,6 @@
 	path = external_libraries/link
 	url = https://github.com/Ableton/link
+[submodule "external_libraries/portaudio"]
+	path = external_libraries/portaudio/portaudio_library
```
@brianlheim what do you think about this directory names? `portaudio` as top level and `portaudio_library` for the actual PA repo? I go back and forth on this, almost thinking that it could be the other way around (`portaudio_library` or `portaudio_something` as top level directory, and `portaudio` for actual PA repo)

brianlheim replied:
is the only reason for the subdirectory the `remove_definitions` call? setting/unsetting the variables can happen in the same scope AFAIK.

dyfer replied:
The main reason was that I'd need to re-set Win32 definitions afterwards and thought moving to a subdirectory would be clearer. 
https://github.com/supercollider/supercollider/blob/c1b26f6458a4303453520f001256340f850996f0/external_libraries/CMakeLists.txt#L35

Also not sure what unsetting C/CXX flags does to the rest of the compilation (though it did work in that commit that I linked). 
Another minor thing is that `asiosdk` can now be in a subdirectory, and not at the same level as other modules, but not sure if that's worth anything.

brianlheim replied:
ok, that makes sense. i'm torn because i liked the directory structure being consistent, but it is a lot easier to understand with portaudio contained inside 2 subdirectories instead of 1. i think let's keep it this way and use the subdirectory names `portaudio/portaudio_submodule`, that's the clearest naming i can think of and it immediately explains the difference between the two. normally, i would suggest `portaudio/portaudio` but not if users are going to be navigating that hierarchy.

brianlheim commented on external_libraries/portaudio/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+
+add_subdirectory("${CMAKE_CURRENT_SOURCE_DIR}/portaudio_library")
+
+if(WIN32)
+    message(STATUS "PortAudio support for ASIO: ${PA_USE_ASIO}")
```
portaudio already prints a message if PA_USE_ASIO is true, so i would suggest this instead:

```suggestion
if(WIN32 AND NOT PA_USE_ASIO)
    message(STATUS "Building Portaudio without support for ASIO")
```

brianlheim commented on lang/CMakeLists.txt:
```diff
@@ -123,7 +123,7 @@ endif()
     add_definitions("-DSC_AUDIO_API_COREAUDIO")
 elseif (AUDIOAPI STREQUAL portaudio)
-    if ((NOT WIN32) OR MSYS)
+    if (SYSTEM_PORTAUDIO)
```
remove the comments on the endif()s below this

brianlheim commented on server/supernova/CMakeLists.txt:
```diff
@@ -145,7 +145,7 @@ endif()
 if(AUDIOAPI STREQUAL jack)
     find_package(Jack REQUIRED)
-elseif(AUDIOAPI STREQUAL portaudio AND ((NOT WIN32) OR MSYS)) # MSYS like Apple
+elseif(AUDIOAPI STREQUAL portaudio AND (SYSTEM_PORTAUDIO))
```
```suggestion
elseif(AUDIOAPI STREQUAL portaudio AND SYSTEM_PORTAUDIO)
```

brianlheim commented on CMakeLists.txt:
```diff
@@ -68,8 +68,7 @@ endif()
   "${CMAKE_COMMAND}" -P "${CMAKE_CURRENT_BINARY_DIR}/cmake_uninstall.cmake")
 
-# identify target architecture, build portaudio before any SC specific
-# definitions break the pa-build
```
this is a noteworthy comment. you don't need to change anything, but we may end up needing to remove more flags as now everything below can potentially be defined when compiling portaudio. that one define is probably the only one. after all we've been building boost libs and yamlcpp this way for quite a while with no issues.

dyfer replied:
Do you mean that we've been building everything with flags set by PortAudio and now we won't?

brianlheim replied:
i mean that now we are building with all the flags set later on in this file (`-DBOOST_DATE_TIME_NO_LIB -DBOOST_ALL_NO_LIB` etc.), where before we weren't

brianlheim reviewed (changes requested):
looking good overall, i just have a few minor comments; once you address those and change portaudio_library to portaudio_submodule i think this can be merged. thanks!

brianlheim commented on server/scsynth/CMakeLists.txt:
```diff
@@ -42,7 +42,7 @@ if(AUDIOAPI STREQUAL jack)
     message(FATAL_ERROR "Jack selected as audio API, but development files not found")
   endif()
-elseif(AUDIOAPI STREQUAL portaudio AND ((NOT WIN32) OR MSYS)) # MSYS like Apple
+elseif(AUDIOAPI STREQUAL portaudio AND (SYSTEM_PORTAUDIO))
```
```suggestion
elseif(AUDIOAPI STREQUAL portaudio AND SYSTEM_PORTAUDIO)
```

brianlheim commented on external_libraries/CMakeLists.txt:
```diff
@@ -6,6 +6,11 @@ if (NOT EXISTS ${CMAKE_CURRENT_SOURCE_DIR}/nova-simd/vec.hpp)
 
+if(NOT SYSTEM_PORTAUDIO)
+    message(STATUS "PortAudio: building static library")
+    add_subdirectory("${CMAKE_SOURCE_DIR}/external_libraries/portaudio")
```
sorry, just noticed this--

```suggestion
    add_subdirectory(portaudio)
```

brianlheim commented on external_libraries/portaudio/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+# disable install target
+set(PA_DISABLE_INSTALL ON CACHE BOOL "Disable PortAudio install" FORCE)
+
+add_subdirectory("${CMAKE_CURRENT_SOURCE_DIR}/portaudio_submodule")
```
```suggestion
add_subdirectory(portaudio_submodule)
```

brianlheim reviewed (changes requested):
thanks! sorry, two small things i just noticed

brianlheim reviewed (approved):
thanks!
